man.vim: implement 'tagfunc'

bobrippling opened the pull request:
This is a follow up PR to #11199, using `'tagfunc'` in `man.vim`, which gives us man page tag jump functionality for `<C-W><C-]>` and `<C-W>}`.

bobrippling committed:
man.vim: use 'tagfunc' instead of remapping

man#pop_tag() is also no longer used

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -20,11 +20,12 @@ setlocal wrap breakindent linebreak
   nnoremap <silent> <buffer> gO         :call man#show_toc()<CR>
-  nnoremap <silent> <buffer> <C-]>      :Man<CR>
   nnoremap <silent> <buffer> K          :Man<CR>
   nnoremap <silent> <buffer> <C-T>      :call man#pop_tag()<CR>
```
should be able to remove `<C-T>` too, I guess.

bobrippling replied:
ah yeah, cheers :+1:

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -406,4 +410,32 @@ function! man#init_pager() abort
+    \   _, name -> {
+    \     "name": name,
+    \     "filename": dummyfile,
+    \     "cmd": "Man " . name " we can't execute :Man like this, sandbox prevents it
```
I assume that restriction is for 'modeline', but now that `modelineexpr=0` is the default,  we could remove the sandbox restriction here. 

Perhaps 'modelineexpr' should be permanent/non-optional.

bobrippling replied:
I've gone with the other approach (which feels less hacky without this dummy filename stuff), but out of interest, by removing the sandbox restriction, do you mean making a change to the C code to disable sandbox execution for `'tagfunc'` ?

justinmk replied:
> , by removing the sandbox restriction, do you mean making a change to the C code to disable sandbox execution for 'tagfunc' ?

Yes

andymass commented on runtime/autoload/man.vim:
```diff
@@ -406,4 +410,32 @@ function! man#init_pager() abort
+    let candidates = s:get_paths(sect, a:pattern)
+    return map(candidates, {
+    \ _, path -> {
+    \     'name': s:extract_sect_and_name_path(path)[1], " this leaves us on the raw man-page file
```
Why doesn't it work to direct to man:// path?  Are the man plugin autocmds not called?

justinmk replied:
(I assume) because of the sandbox

bobrippling replied:
Ah, redirecting to `man://` is exactly what I'm after, thanks! The autocmds are called with this now, there were just sandbox issues for the alternative `:Man` implementation (which I've now ditched)

bobrippling committed:
man.vim: pull out s:get_paths()

bobrippling committed:
man.vim: `:Man` preserves the tag stack

bobrippling commented on runtime/autoload/man.vim:
```diff
@@ -406,4 +419,19 @@ function! man#init_pager() abort
 
+function! man#goto_tag(pattern, flags, info) abort
+  " currently no support for section completion
+  let sect = ""
```
I suppose here we could parse pattern for a section specifier, and if one isn't found, perhaps fallback to `b:man_sect`

bobrippling replied:
Decided to go with this approach, except without falling back to `b:man_sect` - almost all man pages have `man(X)` references, so we'd be unlikely to need to fallback to a section, so I think it would be unnecessary complexity/code for no benefit.

bobrippling committed:
man.vim: parse the section from the tag

bobrippling commented on runtime/autoload/man.vim:
```diff
@@ -64,8 +64,9 @@ function! man#open_page(count, count1, mods, ...) abort
-  call s:push_tag()
-  let bufname = 'man://'.name.(empty(sect)?'':'('.sect.')')
+  let fullname = name.(empty(sect)?'':'('.sect.')')
+  call s:push_tag(fullname)
```
I suppose this is existing behaviour, but you could argue that we don't want to push a tag or touch the tag stack at all now - `'tagfunc'` handles this all for us.

My reasoning is that as a user, would I expect the tag stack to change when I run a `:Man` command? Particularly as we now have `:tag <man page>` available, at least in man-page buffers. I'd be interested to hear your thoughts.

justinmk replied:
Either way is ok. Maybe keep it as is.

andymass replied:
I haven't actually tried it but it should be possible to do

```vim
    let l:page = 'vim'
    let [l:buf, l:save_tfu] = [bufnr(), &tagfunc]
    try
        set tagfunc=ManTagFunc
        execute "tag" l:page 
    finally
        call setbufvar(l:buf, '&tagfunc', l:save_tfu))
    endtry
```

bobrippling replied:
Great - I've gone with that. Everything should be ready now, pending another review round.

bobrippling committed:
man.vim: remove push_tag and simplify man#open_page

bobrippling committed:
Remove eventignore - :Man now uses :tag to populate the page

bobrippling committed:
Don't attempt swapfiles for man pages

This is because we now use :tag to open a man page, which attempts
to open a swap file for a path under man://...

bobrippling committed:
Sort man pages by relevance during goto_tag()

bobrippling commented:
This should now be ready - a few changes were needed still - `'eventignore'` was blocking `:Man` from working (now that `:Man` uses `:tag`) and swapfiles were problematic because of how `:tag` works.

I've also had the tag function sort man pages by relevance, so `:tag ls` will give us `ls(1)` before `lspci(1)`, for example.

blueyed commented:
offtopic: will check/test this later, but just reminded me of that having coverage being reported for Vimscript would be nice (https://github.com/vim/vim/issues/4810).

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -20,13 +20,13 @@ setlocal wrap breakindent linebreak
   nnoremap <silent> <buffer> gO         :call man#show_toc()<CR>
-  nnoremap <silent> <buffer> <C-]>      :Man<CR>
   nnoremap <silent> <buffer> K          :Man<CR>
-  nnoremap <silent> <buffer> <C-T>      :call man#pop_tag()<CR>
```
I suppose `K` should be mapped to `<C-]>` now? 

But I'll leave it as-is, since I noticed a difference: with e.g. `:Man git-notes`, cursor at the caret position:

```
GIT
       Part of the git(1) suite
                        ^
```

`K` (mapped to `:Man<CR>`) follows the `git(1)` reference correctly ,whereas `<c-]>` searches for `suite`.

bobrippling replied:
Yeah that's an interesting one. I think `iskeyword` including `'('` and `')'` when `filetype=man` is the right fix to make for this. I've opened a #11457 to sort this.

I think having `keywordprg=:Man` is okay though, it feels more natural than adding a mapping and ignoring any potential custom `'keywordprg'` a user might have set for man pages. Would you agree?

justinmk replied:
> having keywordprg=:Man is okay though, it feels more natural than adding a mapping

Oh, yes. So the K mapping was  an oversight , or at least it is after 'tagfunc' was implemented.
